File: top_interview_questions/1_easy/1_array/rotate_image/solution.py
from copy import deepcopy
from typing import List


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        size = len(matrix)
        matrix_copy = deepcopy(matrix)

        for i in range(size):
            for j in range(size):
                matrix[j][size-i-1] = matrix_copy[i][j]

File: top_interview_questions/1_easy/1_array/rotate_image/solution.go
package main

import "fmt"

func main() {
	image := [][]int{
		{1, 2, 3, 4},
		{5, 6, 7, 8},
		{9, 10, 11, 12},
		{13, 14, 15, 16},
	}
	rotate(image)
	fmt.Println(image)
}

func rotate(matrix [][]int) {
	size := len(matrix)
	matrixCopy := make([][]int, size)
	copy2d(matrixCopy, matrix)

	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			matrix[j][size-i-1] = matrixCopy[i][j]
		}
	}
}

func copy2d(dst, src [][]int) {
	for i := range src {
		dst[i] = make([]int, len(src[i]))
		copy(dst[i], src[i])
	}
}
